Docs main placeholder pr concept (#1026)
nephio-project/porch: docs/content/en/docs/2_concepts/placeholder-package-revision.md

---
title: "Placeholder Package Revision"
type: docs
weight: 4
description: |

>   The placeholder package revision tracks the configured Git branch for a package. This page explains its behaviour, lifecycle interactions, and how it relates to GitOps workflows.

Reviewer: ```suggestion
  The placeholder package revision tracks the most recent content of a package in its repository. This page explains its behaviour, lifecycle interactions, and how it relates to GitOps workflows.
```

---

## Overview

> The placeholder package revision is a special PackageRevision that Porch creates automatically to represent the state of a package on its configured Git branch. It acts as a "branch-tracking" reference, always reflecting the content of the most recently published revision.

Reviewer: ```suggestion
The placeholder package revision is a special PackageRevision that Porch creates automatically as a reference the latest released version of a package. It acts as a "branch-tracking" reference. The placeholder package revision points at the version of the package revision at the HEAD of the branch of the repository that the package revision is stored on.
```

**Identifying a placeholder:**

| Property | Value |
|----------|-------|
| Revision number | `-1` |
| Workspace name | The branch configured on the Repository CR (`spec.git.branch`), commonly `main` |
| Naming convention | `{repository-name}.{package-name}.{branch-name}` |

> | Lifecycle | `Published` |

Author: what? this makes 0 sense there is no `||` anywhere in this table and no empty columns? this is standard a markdown table.

> | Lifecycle | `Published` |

Author: closing comment as i think this is just copilot hallucinating...

> There is always at most one placeholder package revision per package.

Reviewer: ```suggestion
There is one and only one placeholder package revision per package.
```

## When Is It Created?

> The placeholder is created automatically when the **first revision** of a package is published (transitions from Proposed to Published). You never create it manually.

Reviewer 2: ```suggestion
The placeholder package revision is created automatically when the **first revision** of a package is published (transitions from Proposed to Published). You cannot create it manually.
```

> For example, publishing `example-repository.my-package.v1` (revision 1) will also create `example-repository.my-package.main` (revision -1) with the same content.

Author: fair but low level comment. its easier for a simple example to use numbers to show upgrading/rolling back than words. closing copilot comment.

## When Is It Updated?

Each time a **new revision** of the same package is published, the placeholder is updated to reflect that revision's content and tasks.

For example:
1. Publish v1 (revision 1): placeholder is **created** with v1's content

> 2. Publish v2 (revision 2): placeholder is **updated** with v2's content

Author: fair but low level comment. its easier for a simple example to use numbers to show upgrading/rolling back than words. closing copilot comment

> The placeholder **only moves forward** through explicit publish operations.

Reviewer: ```suggestion
The placeholder **only moves forward** on each explicit publish operations.
```

> ## What Happens When a Revision Is Deleted?

Reviewer: ```suggestion
## What Happens When the Latest Package Revision Is Deleted?
```

> Deleting a published revision (even if it is the latest) does **not** cause the placeholder to roll back to a previous revision. The placeholder retains the content it had at the time of the last publish.

Reviewer: ```suggestion
Deleting the latest published package revision does **not** cause the placeholder to roll back to a previous revision. The placeholder retains the content it had at the time of the last publish. It still refers to the version of the Package Revision at the HEAD of the branch in git.
```

This is intentional. Consider the scenario:

1. v1 is published (placeholder reflects v1)
2. v2 is published (placeholder reflects v2)
3. v2 is deleted

After step 3, the placeholder **still reflects v2's content**. It does not fall back to v1.

> **Why?** Because deletion of a PackageRevision in Porch:

Reviewer 2: ```suggestion
**This occurs due to deletion of a PackageRevision in Porch:
```

Very nitpicky, I know, just not a fan of starting a sentence with Because

- Removes the tag/branch reference from Git
- Removes the Porch metadata

> But it does **not** perform a `git revert` on the tracked branch. The actual content on that branch in Git is unchanged. The placeholder remains consistent with the real Git state.

Reviewer 2: ```suggestion
However, it does **not** perform a `git revert` on the tracked branch. The actual content on that branch in Git is unchanged. The placeholder remains consistent with the real Git state.
```

## How to Roll Back a Package

Since the placeholder only moves forward, the intended way to "roll back" is to publish a new revision with the desired content:

1. Copy the older revision to create a new draft:

   ```bash
   porchctl rpkg copy example-repository.my-package.v1 \
     --namespace=default --workspace=v3
   ```

2. Propose and approve it:

   ```bash
   porchctl rpkg propose example-repository.my-package.v3 --namespace=default
   porchctl rpkg approve example-repository.my-package.v3 --namespace=default
   ```

3. The placeholder now reflects v3 (which has v1's content)

This preserves linear Git history and makes the intent **explicit**.

## Relationship to GitOps

> In a GitOps workflow, a reconciler (such as Flux or ArgoCD) watches a branch in Git for changes. The placeholder package revision corresponds directly to this branch.

Reviewer: ```suggestion
In a GitOps workflow, a reconciler (such as Flux or ArgoCD) watches a branch in Git for changes. The placeholder package revision references this branch.
```

The branch is configured via `spec.git.branch` on the Repository CR. This is commonly `main` but can be any branch name (e.g. `production`, `release`, `staging`). The placeholder's workspace name will match whatever branch is configured.

Because the placeholder mirrors the real state of that branch:

- What the reconciler sees in Git is what the placeholder represents in Porch
- There is no divergence between Porch's view and the deployed state
- Deleting a PackageRevision in Porch does not cause unexpected changes in the deployed environment

## Can the Placeholder Be Deleted?

Yes. The placeholder is **not immutable**. It can be deleted via `porchctl` like any other PackageRevision. This is useful during package cleanup (e.g. removing a package entirely from a repository).

> Deleting the placeholder does **not** affect the underlying Git branch content.

Reviewer 3: Is it not? I thought that deleting the placeholder removes it from the `main` branch.

Author: yes you are right. i was under the impression that the latest tag PR is what decides this and its removal means content in main is removed. tested and it is as you say the main is what dictates this. i will reformat

Author: done

## Summary

| Action | Effect on Placeholder |
|--------|----------------------|
| First revision published | Placeholder **created** |
| Subsequent revision published | Placeholder **updated** to new content |
| Published revision deleted | Placeholder **unchanged** (no rollback) |

> | Placeholder itself deleted | Removed from Porch (Git branch unaffected) |

Reviewer: See @nagygergo comments earlier.

Author: done

> | New revision published after placeholder deletion | Placeholder **recreated** |

Author: same as comment above. closing comment as i think this is just copilot hallucinating...

> | New revision published after placeholder deletion | Placeholder **recreated** |

Reviewer 3: created with contents of the new version

Author: will amend

Author: done

## Restrictions

> The placeholder cannot be used as a source for certain operations:

Reviewer 3: Are these actually enforced?

Author: yes they are i believe i just checked them (at least at the cli level, unsure about api). Haven't checked upgrade yet but clone and copy seem fine

Copy:

```console
catalin@E-5CG2302VDX:~/Nephio/2026/11-test/porch$ porchctl rpkg copy example-repository.demo-package.main --workspace=example-workspace-2 --namespace=example-namespace
Error: Internal error occurred: source revision may not be the placeholder package revision example-repository/example-repository.demo-package.main
```

Clone:

```console
catalin@E-5CG2302VDX:~/Nephio/2026/11-test/porch$ porchctl rpkg clone example-repository.demo-package.main demo-package-2 --repository=test-blueprints --workspace=example-workspace-2 --namespace=example-namespace
Error: Internal error occurred: source revision may not be the placeholder package revision example-repository/example-repository.demo-package.main
```

Reviewer 4: Yes - the checks are at the level of applying the tasks, so they cover the API as well.

- **Clone**: cannot clone from a placeholder
- **Edit/Copy**: cannot edit or copy a placeholder

> - **Upgrade**: cannot upgrade to or from a placeholder

Reviewer 4: ```suggestion
- **Upgrade**: cannot upgrade a placeholder or use a placeholder as the target upstream of an upgrade
```
(using a placeholder as the source upstream is already ruled out by not being able to clone a placeholder)

type: docs
weight: 4
description: |
  The placeholder package revision tracks the most recent content of a package in its repository. This page explains its behaviour, lifecycle interactions, and how it relates to GitOps workflows.
---

## Overview

> The placeholder package revision is a special PackageRevision that Porch creates automatically as a reference the latest released version of a package. It acts as a "branch-tracking" reference. The placeholder package revision points at the version of the package revision at the HEAD of the branch of the repository that the package revision is stored on.

Reviewer 4: ```suggestion
The placeholder package revision is a special PackageRevision that Porch creates automatically as a reference to the latest released version of a package. It acts as a "branch-tracking" reference: the placeholder package revision points at the version of the package at the HEAD of the branch of the repository where the package revision is stored.
```

After step 3, the placeholder **still reflects v2's content**. It does not fall back to v1.

> **This occurs due to deletion of a PackageRevision in Porch:

Reviewer 4: ```suggestion
**This occurs because** the deletion of a PackageRevision in Porch:
```
